File: crates/lanekeep-rules/rules/no-broad-except.ts
import { defineRule } from 'lanekeep'

/**
 * Catching everything in Python.
 *
 * A bare `except:` catches `KeyboardInterrupt` and `SystemExit` too, so it swallows the
 * user pressing Ctrl-C. `except Exception:` is narrower and still catches every bug in the
 * block — a typo'd attribute, a `None` where an object was expected — and reports whatever
 * the handler decided the failure was.
 *
 * The resolver is what makes this precise rather than a text match. A project that defines
 * its own `Exception` class, or imports one, is not catching the builtin, and firing there
 * would be wrong. `ctx.bindingKind` answers that: a name with any local binding is not the
 * builtin.
 *
 * @example
 * ```ts
 * import noBroadExcept from 'lanekeep/no-broad-except'
 *
 * export default defineConfig({ rules: [noBroadExcept] })
 * ```
 */
export default defineRule({
  id: 'lanekeep/no-broad-except',
  language: 'python',
  severity: 'error',

  card: {
    message: 'except catches too much',
    remediation:
      'name the exceptions this block can actually raise, so an unexpected one still surfaces',
    examples: {
      bad: 'try:\n    parse(raw)\nexcept Exception:\n    return None',
      good: 'try:\n    parse(raw)\nexcept ValueError:\n    return None',
    },
  },

  gates: { fileContains: ['except'] },

  query: '(except_clause) @clause',

  check(ctx, m) {
    const children = ctx.namedChildren(m.clause)

    // `except:` — no exception named at all. The first named child of a bare clause is its
    // block, because `except` itself is an anonymous token.
    const caught = children[0]
    if (!caught || ctx.kind(caught) === 'block') {
      ctx.report(m.clause, {
        message: 'bare `except:` also catches KeyboardInterrupt and SystemExit',
      })
      return
    }

    // `except Exception:` and `except Exception as e:`, which wraps the name in a pattern.
    const named =
      ctx.kind(caught) === 'as_pattern' ? ctx.namedChildren(caught)[0] : caught
    if (!named || ctx.kind(named) !== 'identifier') return

    const name = ctx.text(named)
    if (name !== 'Exception' && name !== 'BaseException') return

    // Locally defined or imported under that name — a different exception entirely.
    if (ctx.bindingKind(named)) return

    ctx.report(m.clause, {
      message: `\`except ${name}\` catches every error the block can raise, including bugs`,
    })
  },
})

File: crates/lanekeep-rules/tests/no_broad_except.rs
//! `lanekeep/no-broad-except`, run through the real engine.
//!
//! The first built-in rule for a language other than JavaScript, so these check the whole
//! path as much as the rule: a Python file reaching a Python grammar, and the Python binding
//! resolver answering `ctx.bindingKind` from inside the sandbox.

#![expect(
    clippy::expect_used,
    reason = "`clippy.toml`'s allow-*-in-tests only reaches `#[test]` functions and \
              `#[cfg(test)]` modules. The helpers below are neither, so the grant it \
              already makes for unit tests has to be restated for them."
)]

use lanekeep_testkit::RuleTester;

fn tester() -> RuleTester {
    let source = lanekeep_rules::source("no-broad-except").expect("the rule ships");
    RuleTester::with_extension("no-broad-except", source, "py").expect("builds")
}

#[test]
fn a_named_exception_passes() {
    tester()
        .accepts("try:\n    parse(raw)\nexcept ValueError:\n    return None\n")
        .expect("naming what the block can raise is the point");
}

#[test]
fn several_named_exceptions_pass() {
    tester()
        .accepts("try:\n    go()\nexcept (ValueError, KeyError):\n    pass\n")
        .expect("a tuple of named exceptions is still specific");
}

#[test]
fn a_bare_except_is_reported() {
    tester()
        .reports_at("try:\n    go()\nexcept:\n    pass\n", &[(3, 1)])
        .expect("a bare except also swallows Ctrl-C");
}

#[test]
fn catching_exception_is_reported() {
    tester()
        .reports_at("try:\n    go()\nexcept Exception:\n    pass\n", &[(3, 1)])
        .expect("Exception covers every bug in the block");
}

#[test]
fn catching_base_exception_is_reported() {
    tester()
        .reports_at(
            "try:\n    go()\nexcept BaseException:\n    pass\n",
            &[(3, 1)],
        )
        .expect("BaseException is broader still");
}

#[test]
fn the_as_form_is_reported_too() {
    // `except Exception as e` wraps the name in an `as_pattern`, so a rule reading the
    // clause's first child directly would miss it.
    tester()
        .reports_at(
            "try:\n    go()\nexcept Exception as err:\n    log(err)\n",
            &[(3, 1)],
        )
        .expect("binding the error does not narrow what is caught");
}

#[test]
fn a_project_defined_exception_of_the_same_name_passes() {
    // The reason this rule uses the resolver rather than matching text. A project with its
    // own `Exception` class is not catching the builtin, and reporting it would be wrong.
    tester()
        .accepts("class Exception:\n    pass\n\ntry:\n    go()\nexcept Exception:\n    pass\n")
        .expect("a local class shadows the builtin");
}

#[test]
fn an_imported_exception_of_the_same_name_passes() {
    tester()
        .accepts("from .errors import Exception\n\ntry:\n    go()\nexcept Exception:\n    pass\n")
        .expect("an imported name is not the builtin either");
}
